mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -1195,248 +1204,148 @@ createVerifierFunc(ModuleOp &module, const KernelIF &kernel,
-    } else {
-      maxPercentVal = fval000001; // 0.0001 %
-    }
-
-    // <test> >= <max_percent>
-    cmpVal = testBody.create<arith::CmpFOp>(loc, arith::CmpFPredicate::UGT,
-                                            absfOp, maxPercentVal);
-    if (elemType.getIntOrFloatBitWidth() < 32) {
-      // && <cpu> >= 0.001f
-      auto cmp1Op = testBody.create<arith::CmpFOp>(
-          loc, arith::CmpFPredicate::UGT, absCpuVal, fval001);
-
-      cmpVal = testBody.create<arith::AndIOp>(loc, cmpVal, cmp1Op);
-
-      // TODO?: check for GPU inf
-    }
-
-  } else if (elemType.isF32()) {
-    cmpVal = loopB.create<arith::CmpFOp>(loc, arith::CmpFPredicate::UNE,
-                                         cpuLoadVal, gpuLoadVal);
-    percentDiffVal = loopB.create<arith::SubFOp>(loc, cpuLoadVal, gpuLoadVal);
+  auto getF32Val = [&](float val) -> mlir::Value {
+    llvm::APFloat apVal(val);
+    return b.create<arith::ConstantFloatOp>(loc, apVal, floatType);
+  };
+  // Thresholds for different metrics
+  // RMS: 0.00003f for all data types
+  // absDiff: 100.0f for all data types, i.e. the maxAbsDiff metric is disabled
+  // relDiff 100.0f for f16, i.e. maxRelDiff metric is disabled for f16
+  // datatypes
+  //         0.000001f for other data types
+  auto thr_RMS = getF32Val(RMSThreshold.getValue());
+  auto thr_absDiff = getF32Val(absDiffThreshold.getValue());
+  mlir::Value thr_relDiff;
+  if (testOutType.isF16())
+    thr_relDiff = getF32Val(relDiffThreshold.getValue());
+  else
+    thr_relDiff = getF32Val(0.000001f);
+
+  // obtain function name of the verifier wrapper
+  std::string verifyFuncName = "mcpuVerify5D";
+  if (valElemType.isF32()) {
+    verifyFuncName += "Float";
+  } else if (valElemType.isInteger(32)) {
+    verifyFuncName += "Int32";
   } else {
-    cmpVal = loopB.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne,
-                                         cpuLoadVal, gpuLoadVal);
+    llvm::errs() << "Unsupported type of validation function output: ";
+    valElemType.dump();
+    llvm::errs() << " (Only f32 and int32 are supported)\n";
+    exit(1);
   }
 
-  scf::IfOp ifOp = testBody.create<scf::IfOp>(loc, cmpVal, false);
-  auto thenBody = ifOp.getThenBodyBuilder();
-
-  thenBody.create<memref::StoreOp>(loc, c0ConstantInt32Op, cmpResultAllocOp,
-                                   ValueRange{c0IndexOp});
-
-  // call mcpuPrintF32(float f1, float f2)
-  func::FuncOp printFunc;
-  if (elemType.isIntOrIndex()) {
-    printFunc = makeFuncDecl(module, "mcpuPrintInt32", {intType, intType});
+  auto mr5DUnkType = MemRefType::get({-1, -1, -1, -1, -1}, valElemType);
+  bool isTestAndValSameType =
+      (testOutType.isInteger(32) || testOutType.isF32());
+
+  mlir::Value testResult, valResult; // Values passed to the verify function
+  mlir::Value testResultNew;         // Values used for type conversion
+  if (!isTestAndValSameType) {
+    // When gpu kernel output data type = f16 | bf16, type conversions
```

Comment:
From the validation function perspective, since @yiqian1 has implemented different validation functions ([#738](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/738)), we may have some options:

1. C/C++ implementation: it is much easier to use f32 as the data type
2. MLIR implementation: I am not sure if it is possible to perform the computation in f16. But as a validation function, f32 is generally used to validate f16 computations. Can you confirm, @yiqian1?

From the verification function perspective, since it is implemented in C/C++, f32 is the best option for the computation. Hence the conversion.

Reply:
I will say that, given that MIOpen already has it and that it's a header-only library, it's not inherently unreasonable to pull in `half.hpp` and make this all templatized on datatype. We could probably similarly pull in a bfloat type

Reply:
@zhanglx13 I wouldn't worry too much on having to implement something for this point. It isn't unreasonable for fp16 to be validated in fp32 your and your comments reminded me the reason of it. Thanks.